### PR TITLE
Added Trilu CUDA kernel.

### PR DESCRIPTION
**Pull request by satyajandhyala**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/tensor/trilu.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#include "core/providers/cuda/cuda_kernel.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+class Trilu final : public CudaKernel {
+ public:
+  Trilu(const OpKernelInfo& info) : CudaKernel(info), upper_(true) {
+    int64_t temp_upper;
+    Status status = info.GetAttr<int64_t>("upper", &temp_upper);
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/trilu.h:13]: can use GetAttrOrDefault()
default value for 'upper' is true.

File: onnxruntime/core/providers/cuda/tensor/trilu.cc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+    Trilu,
+    kOnnxDomain,
+    14,
+    kCudaExecutionProvider,
+    (*KernelDefBuilder::Create())
+        .InputMemoryType(OrtMemTypeCPUInput, 1)
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/trilu.cc:19]: can include .MayInplace(0, 0)

File: onnxruntime/core/providers/cuda/tensor/trilu.cc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Input tensor should have a rank of at least 2");
+  }
+  Tensor* output = ctx->Output(0, shape);
+  auto batch_size = input_dims[rank - 1] * input_dims[rank - 2];
+  if (batch_size == 0) {
+    return Status::OK();
+  }
+  TensorPitches input_pitches(input_dims);
+  TArray<int64_t> input_strides(input_pitches);
+  const fast_divmod row_col_divmod_indices(gsl::narrow_cast<int>(input_strides[rank - 1] * input_dims[rank - 1]));
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/trilu.cc:47]: input_strides[rank - 1] is always 1

File: onnxruntime/core/providers/cuda/tensor/trilu.cc [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  if (input_ptr == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
+  const Tensor& input = *input_ptr;
+  const TensorShape& shape = input.Shape();
+  const std::vector<int64_t>& input_dims = shape.GetDims();
+  int32_t rank = gsl::narrow_cast<int32_t>(input_dims.size());
+  if (rank < 2) {
+    return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Input tensor should have a rank of at least 2");
+  }
+  Tensor* output = ctx->Output(0, shape);
+  auto batch_size = input_dims[rank - 1] * input_dims[rank - 2];
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/trilu.cc:41]: 'batch_size' is usually a name reserved for the leading dimensions. 
Consider a different name.

File: onnxruntime/core/providers/cuda/tensor/trilu_impl.cu [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "trilu_impl.h"
+#include <stdio.h>
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__global__ void TriluKernel(
+    bool upper,
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/trilu_impl.cu:12]: bool upper can be a template variable. 
It will save some computation in the kernel.

**Comment by SherlockNoMad** [comment]
looks good to me overall.

File: onnxruntime/core/providers/cuda/tensor/trilu.cc (hunk 4 above)
Comment (satyajandhyala) [reply onnxruntime/core/providers/cuda/tensor/trilu.cc:41]: Okay. I will try to find another name. However pl look at 
["Trilu takes one input tensor of shape [*, N, M], where * is zero or more batch dimensions"](https://github.com/onnx/onnx/blob/master/docs/Operators.md#:~:text=Trilu%20takes%20one%20input%20tensor%20of%20shape%20%5B*%2C%20N%2C%20M%5D%2C%20where%20*%20is%20zero%20or%20more%20batch%20dimensions)
In this case all dimensions except the last two dimensions are considered batch dimensions.

**Comment by ytaous** [comment]
Kindly add some UTS pls?

**Comment by satyajandhyala** [comment]
> Kindly add some UTS pls?
Using forward only UT there is a perf improvement by about 13-14%. The computation is 90ms (master) vs. 80ms
image

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 6]
```diff
@@ -96,6 +96,11 @@ IMPLEMENT_GRADIENT_BUILDER(GetTanhGradient) {
               {GI(0)})};
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetTriluGradient) {
+  return std::vector<NodeDef>{
+      NodeDef(OpDef{"Trilu", kMSDomain, 1}, {GO(0)}, {GI(0)})};
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:101]: the input should be {GO(0), I(1)}, as k is also critical.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 7]
```diff
@@ -96,6 +96,11 @@ IMPLEMENT_GRADIENT_BUILDER(GetTanhGradient) {
               {GI(0)})};
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetTriluGradient) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:99]: As we have added gradient builder, we should also add tests in gradient_ops_test.cc

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 8]
```diff
@@ -4320,3 +4320,28 @@ def test_serialize_ortmodule():
     assert ort_out is not None
     _test_helpers.assert_values_are_close(pt_out, ort_out)
     _test_helpers.assert_gradients_match_and_reset_gradient(ort_model_2, pt_model)
+
+def test_trilu():
+    class SimpleTriuModel(torch.nn.Module):
+        def forward(self, x):
+            y = torch.triu(x)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:4327]: need test case for different k value?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 9]
```diff
@@ -4320,3 +4320,28 @@ def test_serialize_ortmodule():
...
+        loss = prediction.sum()
+        loss.backward()
+        return prediction, loss
+
+    device = 'cuda'
+    pt_model = SimpleTriuModel().to(device)
+    ort_model = ORTModule(copy.deepcopy(pt_model))
+
+    for i in range(10):
+        pt_x = torch.rand((1, 10, 10), requires_grad=True, device=device)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:4341]: need more test cases for various input shapes.

File: onnxruntime/core/providers/cuda/tensor/trilu.cc [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+  size_t element_size = input.DataType()->Size();
+  return TriluImpl(
+      this->Stream(),
+      upper_,
+      element_size,
+      k_val,
+      input.DataRaw(),
+      output->MutableDataRaw(),
+      gsl::narrow<int>(output->Shape().Size()),
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/trilu.cc:57]: shape.Size() is sufficient?

File: onnxruntime/core/providers/cuda/tensor/trilu_impl.cu [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    T* output_data,
+    const CUDA_LONG N,
+    const fast_divmod batch_divmod_indices,
+    const fast_divmod row_col_divmod_indices) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
+
+  int row, col;
+
+  row_col_divmod_indices.divmod(batch_divmod_indices.mod(id), row, col);
+  output_data[id] = ((((row + k) <= col) && upper) || (((row + k) >= col) && !upper)) ? input_data[id] : 0;
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/trilu_impl.cu:23]: I am not sure if we need to rewrite into following pattern so that the compiler can optimize the bool-type template parameter. 

if (upper) 
...
else
...

**Review by SherlockNoMad — APPROVED** [review]
